Fast and Flexible GPU Quantization for Transformers.md

```diff
@@ -131,10 +131,72 @@ By zooming in (4 rows, 16 columns) we hopefully get a clearer picture of the lay
 #### Column Turing (CUBLASLT_ORDER_COL4_4R2_8C)
 ![](_attachments/Pasted%20image%2020230329113436.png)
 
-### Column Ampere (CUBLASLT_ORDER_COL32_2R_4R4)
+#### Column Ampere (CUBLASLT_ORDER_COL32_2R_4R4)
 ![](_attachments/Pasted%20image%2020230329113458.png)
 
-While `COL32` might be the most performant option, there exists a tension whereby the cost of the layout conversion may cancel out any gains, and so must either make a design decision à la FasterTransformer and persist the data in the required format, or hide the cost via kernel fusion. The latter approach is similar to how quantize/dequantize overhead is typically hidden, which we will discuss next.
+While `COL32` might be the most performant option, there exists a tension whereby the cost of the layout conversion may cancel out any gains from the reduced precision matmul. Therefore we must either make a design decision à la FasterTransformer and persist the data in the required format, or hide the cost via kernel fusion. The latter approach is similar to how quantize/dequantize overhead is typically hidden, which we will discuss next.
+
+
+### INT8 GEMM Benchmarking
+
+We now look at some peformance numbers for the various flavours of INT8 GEMM. For these benchmarks we wrap the C++ APIs for cuBLASLt and CUTLASS as PyTorch extensions. A detailed guide to timing CUDA kernels with PyTorch can be found [here](https://www.speechmatics.com/company/articles-and-news/timing-operations-in-pytorch). Benchmarks were run on a T4 GPU with an input tensors of shape [2048, 1920] and [1920, 1920]. Whilst mileage may vary for different input shapes, we found the following conclusions to be consistent over a variety of shapes/sizes.
+
+####  INT8 vs. INT32 output precision
+
+$$D=\alpha AB+\beta C$$
+
+One important factor which detemines the performance of an INT8 GEMM (formula above) is the required output type. The matrix multiplication will always have INT dtype for matrices A and B, which then accumulate the outputs into INT32 within the kernel,  but we need to decide whether output matrix C should be INT8 or INT32. 
+
+INT32 return type will be slower as four times as much data is written out (and read into the next kernel). We will also have to dequantize after the matmul to return to FP16. In comparison INT8 return type is faster, but there is a trade-off as accuracy will be impacted as we need to re-quantize the output from INT32 to INT8 within the kernel. More information on this can be found {earlier in the blog}. If the next operation requires FP16 input we will also have to dequantize, however if we require INT8 for the next kernel an INT8 output type can be ideal.
```

Comment:
```suggestion
INT32 return type will be slower as four times as much data is written out (and read into the next kernel). We will also have to dequantize after the matmul to return to FP16. In comparison, INT8 return type is faster but there is a trade-off: accuracy will be made worse, as we need to requantize the output from INT32 to INT8 within the kernel. More information on this can be found {earlier in the blog}. If the next operation requires FP16 input we will also have to dequantize. However, if we require INT8 for the next kernel an INT8 output type can be ideal.
```